smac/intensification/successive_halving.py

```diff
@@ -0,0 +1,462 @@
+        self.max_sh_iter = np.floor(np.log(self.max_budget / self.min_budget) / np.log(eta))
+        self.initial_challengers = int(np.round(self.eta ** (self.max_sh_iter + 1)))
+
+        # for adaptive capping
+        if self.budget_type == 'instance' and self.instance_order != 'budget_random' and self.run_obj_time:
+            self.adaptive_capping = True
+        else:
+            self.adaptive_capping = False
+        self.adaptive_capping_slackfactor = adaptive_capping_slackfactor
+
+    def intensify(self, challengers: typing.List[Configuration],
+                  incumbent: typing.Optional[Configuration],
+                  run_history: RunHistory,
+                  aggregate_func: typing.Callable,
+                  time_bound: float = float(MAXINT),
+                  log_traj: bool = True):
+        """
+        Running intensification via successive halving to determine the incumbent configuration.
+        *Side effect:* adds runs to run_history
+
+        Implementation of successive halving by Jamieson & Talwalkar (2016)
+
+        Parameters
+        ----------
+        challengers : typing.List[Configuration]
+            promising configurations
+        incumbent : Configuration
+            best configuration so far
+        run_history : RunHistory
+            stores all runs we ran so far
+        aggregate_func: typing.Callable
+            aggregate error across instances
+        time_bound : float, optional (default=2 ** 31 - 1)
+            time in [sec] available to perform intensify
+        log_traj: bool
+            whether to log changes of incumbents in trajectory
+
+        Returns
+        -------
+        incumbent: Configuration()
+            current (maybe new) incumbent configuration
+        inc_perf: float
+            empirical performance of incumbent configuration
+        """
+        self.start_time = time.time()
+        self._ta_time = 0
+        self._chall_indx = 0
+        self._num_run = 0
+
+        if time_bound < self._min_time:
+            raise ValueError("time_bound must be >= %f" % self._min_time)
+
+        if isinstance(challengers, ChallengerList):
+            challengers = challengers.challengers
+
+        # selecting first 'n' challengers for the current run, based on eta
+        curr_challengers = challengers[:self.initial_challengers]
+        # challengers can repeated only if optimizing across multiple seeds
+        if self.n_seeds is None or self.n_seeds == 1:
+            curr_challengers = set(curr_challengers) - set(run_history.get_all_configs())
+
+        # randomize instances per successive halving run, if user specifies
+        all_instances = self.instances
+        if self.instance_order == 'budget_random':
+            np.random.shuffle(all_instances)
+
+        # calculating the incumbent's performance for adaptive capping
+        # this check is required because there is no initial run to get the incumbent performance
```

Comment:
The initial design involves 1 full SH run with default + random configurations. Since it is essentially a `intensify()` call from there, that will not have an incumbent.

Reply:
Interesting, I believe this needs to be documented.